src/java_bytecode/generate_java_generic_method.cpp

```diff
@@ -0,0 +1,291 @@
+  {
+    // We should never hit this because the first thing
+    // we do before starting to specialize the method
+    // is check that the proposed symbol name doesn't
+    // already exist.
+    throw unsupported_java_method_specialization_exceptiont(
+      "specialized method symbol already exists");
+  }
+
+  return *symbol_table.lookup(specialized_method.name);
+}
+
+
+/// Given a mapping from generic type variables to concrete types, generate
+/// a string suitable for decorating symbol names with.
+/// \param concrete_types The mapping from type variables to concrete types
+/// \return A string suitable for inserting into a java signature
+///
+/// As an example, if a mapping such as this (pseudo code...):
+/// [ {T,java.lang.Integer}, {U,java.lang.Double} ] is passed in, the return
+/// string would look like "<java.lang.Integer,java.lang,Double>"
+const std::string generate_java_generic_methodt::instantiation_decoration(
+  type_variable_instantiationst concrete_types)
+{
+  std::ostringstream decorated_signature_buffer;
+  decorated_signature_buffer << "<";
+  bool first=true;
+  for(auto &concrete_type_entry : concrete_types)
+  {
+    if(first)
+      first=false;
+    else
+      decorated_signature_buffer << ",";
+
+    decorated_signature_buffer <<
+      concrete_type_entry.second.get(ID_identifier);
+  }
+  decorated_signature_buffer << ">";
+  return decorated_signature_buffer.str();
+}
+
+/// Instantiate a generic type with concrete types using a supplied mapping
+/// from type variable to concrete type.
+/// \param generic_type The typet that should be modified and instantiated.
+/// \param concrete_type_map The mapping from type variable to concrete types.
+void generate_java_generic_methodt::instantiate_generic_types(
+  typet &generic_type,
+  const type_variable_instantiation_mapt
+    &concrete_type_map)
+{
+  if(is_java_generic_parameter(generic_type))
+  {
+    // Instantiate a generic type variable 'T'
+    instantiate_java_generic_parameter(
+      to_java_generic_parameter(generic_type),
+      concrete_type_map);
+  }
+  else if(is_java_generic_type(generic_type))
+  {
+    // Instantiate a generic type, such as 'List<T>'
+    instantiate_java_generic_type(
+      to_java_generic_type(generic_type),
+      concrete_type_map);
+  }
+}
+
+/// Instantiate a given java generic parameter using a supplied mapping from
+/// type variable to concrete type.
+/// \param generic_parameter The java_generic_parametert that should be
+/// modified and instantiated.
+/// \param concrete_types The mapping of type variables to concrete types.
+void generate_java_generic_methodt::instantiate_java_generic_parameter(
+  java_generic_parametert &generic_parameter,
+  const type_variable_instantiation_mapt
+    &concrete_types)
+{
+  INVARIANT(generic_parameter.id()==ID_pointer,
+            "All generic parameters should be pointers in java");
+  INVARIANT(generic_parameter.subtype().id()==ID_symbol,
```

Comment:
Let's discuss this tomorrow - I worry updating this as suggested to `ns.follow(generic_type.subtype()).id()==ID_struct` will break your unit tests